Add support for Git external sources

mateusrodrigues opened the pull request:
This PR supersedes #5.

This introduces the concept of a "Flamenco file", called `flamenco.json`. It will be handled just like any other file in terms of package and series specifiers, but, instead of being copied directly to the destination directory, it will be handled and acted upon according to its type.

The first type supported by this PR is `external_source`, which describes where to get the directory's content. As an example, this is what a Git external source descriptor would look like:

```JSON
//src/eng/test-runner/flamenco.json
{
    "type": "external_source",
    "sourceType": "git",
    "repository": "https://github.com/canonical/dotnet-test-runner",
    "commitish": "v1.2.0",
    "rootDirectory": "src/main",
    "postClone": [
        "make GIT_COMMIT_ID",
        "make GIT_TAG_VERSION"
    ],
    "ignoredFiles": [
        ".gitignore",
        ".github/workflows/build.yml"
    ]
}
```
Where:
- `rootDirectory` changes the directory to be copied to the destination.
- `postClone` specifies terminal commands to run after clone, before deleting the `.git` directory.
- `ignoredFiles` are files removed from the repository before copying it to the destination.

Note that, so far, only `git` is implemented.

There is an AI-generated documentation page on external sources available at [docs/external-sources.md](https://github.com/canonical/flamenco/blob/feat/external-sources-2/docs/external-sources.md).

Closes #3

mateusrodrigues committed:
feat: support Git external source file descriptor

mateusrodrigues committed:
chore: handle invalid descriptor data error

dviererbe commented on src/Flamenco.Packaging/ExternalSources/ExternalSourceBase.cs:
```diff
@@ -0,0 +1,31 @@
+
+public abstract class ExternalSourceBase
+{
+    public abstract Task Download(string destinationDirectory, CancellationToken cancellationToken = default);
```
**issue:** The return type of the only implementation (`GitExternalSource`) is `Task<Result>`. This makes the `Result` unusable.

```suggestion
    public abstract Task<Result> Download(string destinationDirectory, CancellationToken cancellationToken = default);
```

dviererbe commented on src/Flamenco.Packaging/ExternalSources/ExternalSourceBase.cs:
```diff
@@ -0,0 +1,31 @@
+{
+    public abstract Task Download(string destinationDirectory, CancellationToken cancellationToken = default);
+
+    public static ExternalSourceBase Create(FileInfo fileInfo)
```
**todo:** The result of this method should be a `Task<Result<ExternalSourceBase>>`, return an error result instead of throwing exceptions and using the async methods to read the file.

**nitpick:** I think the method should be named `Parse` (or `ParseAsync` if made async), because that is what it does.

mateusrodrigues replied:
I made use of the annotation's metadata field to store which fields were invalid for the operation to fail. While at it, I added [this](https://github.com/canonical/flamenco/pull/9/commits/2143a7637bfd6398bcbfaa45a352b9d9a7e41857) to the Log file to make sure the metadata is print out. Let me know if this is an issue.

dviererbe commented on src/Flamenco.Packaging/ExternalSources/GitExternalSource.cs:
```diff
@@ -0,0 +1,25 @@
+    public string Repository { get; } = repository;
+    public string? Commitish { get; } = commitish;
+
+    public override async Task<Result> Download(string destinationDirectory, CancellationToken cancellationToken = default)
```
**issue:** If you return a `Result`. Note that the calling methods expect that no exception is thrown (except for the `OperationCanceledException`). If any of the git cloning or delete operations fail, it may cause side effects to other operations.

dviererbe commented on src/Flamenco.Packaging/DebianSourcePackageBuilder.cs:
```diff
@@ -337,12 +344,13 @@ private Result<ImmutableDictionary<string, FileInfo>> IndexSourceFiles(
             else
             {
-                return result.WithValue((file.Name, BuildTargetSpecificity: 0));
+                flamencoFile = file.Name.Equals(FlamencoFileName, StringComparison.CurrentCultureIgnoreCase);
```
**question:** Why is the string comparison culture dependent? The name of the file should be independent from the current culture setting of the host where flamenco runs on to have deterministic results. Consider using ordinal comparison.

dviererbe commented on src/Flamenco.Packaging/DebianSourcePackageBuilder.cs:
```diff
@@ -376,6 +384,13 @@ private Result<ImmutableDictionary<string, FileInfo>> IndexSourceFiles(
+        // A specificity of -1 means that the file is not specific to current build target.
+        // A specificity of 0 means that the file is not specific to a source package or series.
+        // A specificity of 1 means that the file is specific to either a source package or a series.
+        // A specificity of 2 means that the file is specific to both a source package and a series.
```
**todo:** We should add constants somewhere. The fact that you need to add a comment shows that this are magic numbers.

dviererbe commented on src/Flamenco.Packaging/FlamencoFileInfo.cs:
```diff
@@ -0,0 +1,7 @@
+
+internal sealed class FlamencoFileInfo(bool flamencoFile, FileInfo fileInfo)
+{
+    public bool FlamencoFile { get; } = flamencoFile;
```
**nitpick:** This should be named `IsFlamencoFile` to indicate that it holds a bool and not something like `FileInfo`.

```suggestion
    public bool IsFlamencoFile { get; } = flamencoFile;
```

dviererbe reviewed (changes requested):
Finishing the review tomorrow, but this is my initial feedback.

**todo:** The downloaded git repositories should be cached (e.g. in the `dist` folder). Otherwise it will download a git repo for every series x package, which makes iterating slower and is effectively like a very small Denial of Service attack on GitHub.

mateusrodrigues commented on src/Flamenco.Packaging/ExternalSources/ExternalSourceBase.cs:
```diff
@@ -0,0 +1,31 @@
+            throw new FileNotFoundException("Descriptor file not found.", fileInfo.FullName);
+        }
+
+        var jsonNode = JsonNode.Parse(File.ReadAllText(fileInfo.FullName));
```
Changed the `Create` method to take a `JsonNode` object instead.

mateusrodrigues committed:
refactor: handle exception during Git clone operation

mateusrodrigues committed:
refactor(ExternalSourceBase): replace exceptions with results

mateusrodrigues committed:
refactor: better naming for variables and constants

mateusrodrigues committed:
chore(Log): print out metadata when it exists

mateusrodrigues committed:
chore: rename ExternalSourceBase.Create to Parse

mateusrodrigues committed:
feat: cache Git repository to avoid multiple clones

This code makes use of a dictionary of SemaphoreSlim
objects to orchestrate the cloning of individual
repositories.

While a repository is cloning, all operations for that
repository are blocked. When finished, all subsequent
operations on that repository will be copies from the
cache directory to its final destination.

This will prevent the same repository being cloned
`release` x `package` times.

mateusrodrigues committed:
chore: better error handling

mateusrodrigues commented on src/Flamenco.Packaging/DebianSourcePackageBuilder.cs:
```diff
@@ -481,14 +499,51 @@ private async Task<Result> CopySourceFilesToDestinationDirectoryAsync(
+        switch (type)
+        {
+            case "external_source":
+                var cacheDirectory = Directory.CreateDirectory(Path.Combine(DestinationDirectory.FullName, "cache"));
```
@dviererbe any thoughts on this? Maybe keeping a flamenco cache within `/var/tmp` instead of `dist/` makes more sense.

mateusrodrigues commented on src/Flamenco.Packaging/ExternalSources/GitExternalSource.cs:
```diff
@@ -0,0 +1,101 @@
+        if (cancellationToken.IsCancellationRequested) return result.WithAnnotation(new OperationCanceled());
+
+        var repositoryCacheDir =
+            new DirectoryInfo(Path.Join(cacheDirectory.FullName, "git", destinationDirectory.Name));
```
The cache key is actually `repositoryCacheDir.FullName`, which is expected to have a single source. A collision here is highly unlikely, if not impossible.

mateusrodrigues committed:
fix: missing jsonNode attribution

mateusrodrigues committed:
feat: add support for post-clone commands

mateusrodrigues commented on src/Flamenco.Packaging/ExternalSources/GitExternalSource.cs:
```diff
@@ -0,0 +1,180 @@
+            var process = new Process
+            {
+                StartInfo = new ProcessStartInfo(
+                    fileName: "/usr/bin/env",
```
We're not expecting to support Windows, so this is a non-issue.

mateusrodrigues committed:
fix: better support for commands with spaces

mateusrodrigues committed:
refactor: remove no-op Result merge

mateusrodrigues commented on src/Flamenco.Packaging/DebianSourcePackageBuilder.cs:
```diff
@@ -481,14 +500,60 @@ private async Task<Result> CopySourceFilesToDestinationDirectoryAsync(
+        switch (type)
+        {
+            case "external_source":
+                var cacheDirectory = Directory.CreateDirectory(Path.Combine(DestinationDirectory.FullName, "cache"));
```
This is addressed in the comment above. Marking this conversation as resolved.

mateusrodrigues commented on src/Flamenco.Packaging/ExternalSources/GitExternalSource.cs:
```diff
@@ -0,0 +1,180 @@
+
+            if (process.ExitCode != 0)
+            {
+                return result.Merge(new Result().WithAnnotation(new PostCloneCommandFailed(command, process.ExitCode)));
```
We are not redirecting the Standard Output and Standard Error, so they are being printed out in the console. This should present enough information for debugging. If not, we can reconsider this.

mateusrodrigues committed:
feat: add support for ignored files in Git external source

The `ignoredFiles` JSON key will contain a list of files in the repository that should be deleted after cloning. This is useful for scenarios where the user does not wish to keep certain files around when cloning a repository into a Debian package, e.g. .gitignore files.

Files that are listed and do not exist in the repository are ignored.

mateusrodrigues committed:
feat: also support directories in ignoredFiles

mateusrodrigues committed:
docs: add external sources documentation

mateusrodrigues committed:
chore: add 'git' and 'make' as dependencies

mateusrodrigues committed:
feat: add support for 'rootDirectory' in Git sources

mateusrodrigues committed:
docs: update external sources docs with 'rootDirectory' changes

mateusrodrigues committed:
feat: use committish in git repo cache dir name

mateusrodrigues committed:
chore: handle process start failure

mateusrodrigues committed:
fix(ci): use snapcraft from stable channel

This avoids the upcoming breaking change that will require snapcraft pack runs to explicitly include 'pack', which is not yet supported by the snapcore/action-build action.

mateusrodrigues committed:
feat: allow source packages without patches

   Processing source packages without patches should only produce a warning to the user saying no d/p/series was found. The resulting source package will not contain a d/p directory at all.

Rinconjr committed:
fix(packaging): use core24 base tar to resolve extraction permission denied

The snap-packaged Flamenco staged its own `/usr/bin/tar` via
`stage-packages`. Inside the snap runtime this staged tar binary cannot
create directories, producing errors such as:

    tar: test/...: Cannot mkdir: Permission denied
    tar: test/...: Cannot open: Permission denied

This broke both Flamenco's orig-tarball extraction and
`dpkg-buildpackage`, which calls `tar` through PATH.

Changes:
- Invoke `/usr/bin/tar` directly in `TarSystemCommand` instead of
  `/usr/bin/env tar`, which resolved to the broken staged binary.
- Remove `tar` from `stage-packages` so `dpkg-buildpackage` also falls
  back to the working `tar` provided by the `core24` base snap.

After this fix `flamenco pack` successfully extracts the orig tarball
and builds the source package in both devmode and strict confinement.
